### PR TITLE
Solution for CSES-2101

### PR DESCRIPTION
**Pull request by dutinmeow**
I'm not sure if my code's style is good enough.

### CODE REVIEW

File: solutions/cses-2101 [hunk 1]
```diff
@@ -0,0 +1,166 @@
+---
+id: cses-2101
+source: CSES
+title: New Roads Queries
+author: Dustin Miao
+---
+
+Let's construct a graph where edges have weights equal to the time they are added. The problem boils down to efficiently finding the path with the minimum maximum edge for each query. 
+
+Let's consider a single query between nodes $u, v$. One way of approaching this is to sort the edge weights from smallest to largest, and then add edges one by one until $u$ and $v$ are connected. If we use a [disjoint set union](https://usaco.guide/gold/dsu), the time complexity is $\mathcal{O}(m \cdot \alpha(n))$, which is too slow.
```
Comment (dongliu0426) [inline solutions/cses-2101:10]:
```suggestion
Let's consider a single query between nodes $u, v$. One way of approaching this is to sort the edge weights from smallest to largest, and then add edges one by one until $u$ and $v$ are connected. If we use a [disjoint set union](/gold/dsu), the time complexity for each query is $\mathcal{O}(m \cdot \alpha(n))$, which is too slow.
```

File: solutions/cses-2101 [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+However, this hints at a faster solution. Note that our slow approach bears many similarities with [kruskal's MST algorithm](https://usaco.guide/gold/mst/#kruskals), which implies that the path between $u$ and $v$ on any minimum spanning tree will minimize the maximum edge weight. For a more formal proof, see [here](https://www.geeksforgeeks.org/minimum-bottleneck-spanning-treembst/).
```
Comment (dongliu0426) [inline solutions/cses-2101:12]:
```suggestion
However, this hints at a faster solution. Note that our slow approach bears many similarities with [Kruskal's MST algorithm](/mst/#kruskals), which implies that the path between $u$ and $v$ on any minimum spanning tree will minimize the maximum edge weight. For a more formal proof, see [here](https://www.geeksforgeeks.org/minimum-bottleneck-spanning-treembst/). 
```

File: solutions/cses-2101 [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+author: Dustin Miao
+---
+
+Let's construct a graph where edges have weights equal to the time they are added. The problem boils down to efficiently finding the path with the minimum maximum edge for each query. 
+
+Let's consider a single query between nodes $u, v$. One way of approaching this is to sort the edge weights from smallest to largest, and then add edges one by one until $u$ and $v$ are connected. If we use a [disjoint set union](https://usaco.guide/gold/dsu), the time complexity is $\mathcal{O}(m \cdot \alpha(n))$, which is too slow.
+
+However, this hints at a faster solution. Note that our slow approach bears many similarities with [kruskal's MST algorithm](https://usaco.guide/gold/mst/#kruskals), which implies that the path between $u$ and $v$ on any minimum spanning tree will minimize the maximum edge weight. For a more formal proof, see [here](https://www.geeksforgeeks.org/minimum-bottleneck-spanning-treembst/). 
+
+All that is left to do is to efficiently query for the minimum node on a tree path. There are many methods to do this, but the simplest is with [binary jumping](https://usaco.guide/plat/binary-jump).
```
Comment (dongliu0426) [inline solutions/cses-2101:14]: There's actually an easier method (modifying the DSU & small to large merging), but wtvr

**Review by dongliu0426 — CHANGES_REQUESTED** [review]
rename file to `cses-2101.mdx`

File: solutions/cses-2101.mdx [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+ 
+/** Disjoint Set Union with Small-to-Large Merging 
+ *  Verified: https://judge.yosupo.jp/problem/unionfind */
```
Comment (SansPapyrus683) [inline solutions/cses-2101.mdx:29]:
```suggestion
// Simple DSU class
```

File: solutions/cses-2101.mdx [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+using namespace std;
+ 
+/** Disjoint Set Union with Small-to-Large Merging 
+ *  Verified: https://judge.yosupo.jp/problem/unionfind */ 
+template<size_t N> struct UnionFind {
+	int par[N], sze[N], max_size;
+ 
+	UnionFind(int n = N) { init(n); }
+ 
+	/** Complexity: O(n) */
```
Comment (SansPapyrus683) [inline solutions/cses-2101.mdx:35]:
```suggestion
```

File: solutions/cses-2101.mdx [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+ 
+/** Disjoint Set Union with Small-to-Large Merging 
+ *  Verified: https://judge.yosupo.jp/problem/unionfind */ 
+template<size_t N> struct UnionFind {
+	int par[N], sze[N], max_size;
+ 
+	UnionFind(int n = N) { init(n); }
```
Comment (SansPapyrus683) [inline solutions/cses-2101.mdx:33]:
```suggestion
	UnionFind(int n=N) { init(n); }
```
think this is the style for default arguments

File: solutions/cses-2101.mdx [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+ 
+	/** Complexity: O(n) */ 
+	void init(int n = N) {
+		iota(par, par + n, 0);
+		fill(sze, sze + n, 1);
+		max_size = 1;
+	}
+ 
+	/** Complexity: O(alpha(n)) */
```
Comment (SansPapyrus683) [inline solutions/cses-2101.mdx:42]: you can just remove all these complexity lines, don't think they're needed

File: solutions/cses-2101.mdx [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+ 
+	/** Complexity: O(alpha(n)) */ 
+	int size(int a) { 
+		return sze[find(a)]; 
+	}
+};
+ 
+int N, M, Q;
+vector<pair<int, int>> G[200005];
+UnionFind<200005> dsu;
```
Comment (SansPapyrus683) [inline solutions/cses-2101.mdx:76]: so magic numbers are bad blah blah blah
put 200k in a number called `MAX_N` or something and then in declaring these arrays
do something like `int arr[MAX_N + 1]`

File: solutions/cses-2101.mdx [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+ 
+namespace LCA {
+	int dep[200005], par[200005][18], val[200005][18];
```
Comment (SansPapyrus683) [inline solutions/cses-2101.mdx:79]: yeah same here lol
also the heck is `18` for

File: solutions/cses-2101.mdx [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+		ans = max(ans, val[a][0]);
+		ans = max(ans, val[b][0]);
+ 
+		return ans;
+	}
+}
+ 
+ 
+int main() {
+	ios_base::sync_with_stdio(0); cin.tie(0);
```
Comment (SansPapyrus683) [inline solutions/cses-2101.mdx:141]: if it runs w/o this, delete it

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
this is what i've nitpicked so far

File: solutions/cses-2101.mdx (hunk 9 above)
Comment (dutinmeow) [reply solutions/cses-2101.mdx]: 18 = log_2 N

Comment (SansPapyrus683) [reply solutions/cses-2101.mdx]: you should definitely add that as a comment

File: solutions/cses-2101.mdx [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+	
+const int MAXN = 2e5;
+const int LOGN = 18;
```
Comment (SansPapyrus683) [inline solutions/cses-2101.mdx:29]: still add a comment for this

Comment (dutinmeow) [reply solutions/cses-2101.mdx:29]: why?????

Comment (SansPapyrus683) [reply solutions/cses-2101.mdx:29]: log what? base what?
just say like `// log of MAXN base 2`

**Review by SansPapyrus683 — APPROVED** [review]
i have arrays but there's nothing i can do

**Pull request closed without merging**